core/block/editor/userdataobject/userdata.go

```diff
@@ -0,0 +1,271 @@
+package userdataobject
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"time"
+
+	anystore "github.com/anyproto/any-store"
+	"github.com/anyproto/any-store/anyenc"
+	"github.com/anyproto/any-sync/util/crypto"
+	"golang.org/x/exp/slices"
+
+	"github.com/anyproto/anytype-heart/core/block/cache"
+	"github.com/anyproto/anytype-heart/core/block/editor/smartblock"
+	"github.com/anyproto/anytype-heart/core/block/editor/storestate"
+	"github.com/anyproto/anytype-heart/core/block/source"
+	"github.com/anyproto/anytype-heart/core/domain"
+	"github.com/anyproto/anytype-heart/pb"
+	"github.com/anyproto/anytype-heart/pkg/lib/bundle"
+	"github.com/anyproto/anytype-heart/pkg/lib/logging"
+	"github.com/anyproto/anytype-heart/pkg/lib/pb/model"
+)
+
+var log = logging.Logger("core.block.editor.userdata")
+
+type UserDataObject interface {
+	smartblock.SmartBlock
+
+	SaveContact(ctx context.Context, profile *model.IdentityProfile, key crypto.SymKey) error
+	DeleteContact(ctx context.Context, identity string) error
+	UpdateContactByDetails(ctx context.Context, id string, details *domain.Details) error
+	UpdateContactByIdentity(ctx context.Context, profile *model.IdentityProfile) (err error)
+	ListContacts(ctx context.Context) ([]*Contact, error)
+}
+
+type userDataObject struct {
+	smartblock.SmartBlock
+	state       *storestate.StoreState
+	storeSource source.Store
+	crdtDb      anystore.DB
+	arenaPool   *anyenc.ArenaPool
+
+	objectCache      cache.ObjectGetter
+	ctx              context.Context
+	cancel           context.CancelFunc
+	onUpdateCallback func()
+}
+
+func New(sb smartblock.SmartBlock, crdtDb anystore.DB, objectCache cache.ObjectGetter) UserDataObject {
+	u := &userDataObject{
+		SmartBlock:  sb,
+		crdtDb:      crdtDb,
+		arenaPool:   &anyenc.ArenaPool{},
+		objectCache: objectCache,
+	}
+	u.onUpdateCallback = u.onUpdate
+	return u
+}
+
+func (u *userDataObject) Init(ctx *smartblock.InitContext) error {
+	u.ctx, u.cancel = context.WithCancel(context.Background())
+	err := u.SmartBlock.Init(ctx)
+	if err != nil {
+		return err
+	}
+	stateStore, err := storestate.New(ctx.Ctx, u.Id(), u.crdtDb, contactsHandler{})
+	if err != nil {
+		return fmt.Errorf("create state store: %w", err)
+	}
+	u.state = stateStore
+
+	storeSource, ok := ctx.Source.(source.Store)
+	if !ok {
+		return fmt.Errorf("source is not store")
+	}
+	u.storeSource = storeSource
+	err = storeSource.ReadStoreDoc(ctx.Ctx, stateStore, u.onUpdateCallback)
+	if err != nil {
+		return fmt.Errorf("read store doc: %w", err)
+	}
+	go u.onUpdateCallback()
+	return nil
+}
+
+func (u *userDataObject) onUpdate() {
+	contacts, err := u.ListContacts(u.ctx)
+	if err != nil {
+		log.Errorf("list contacts: %v", err)
+		return
+	}
+	for _, contact := range contacts {
+		u.createContactAndUpdateDetails(contact)
+	}
+}
+
+func (u *userDataObject) createContactAndUpdateDetails(contact *Contact) {
+	err := cache.DoContextFullID(u.objectCache, u.ctx, domain.FullID{
```

Comment:
Pls move all this logic to a ContactService. This object should not now anything about how the contacts are being created

Reply:
It is not a responsibility of this object